metabase/metabase
# fix(sdk): fix React 17 compatibility with Table visualization

**_Problem_**

For a client using the Embedding SDK with React 17, they were running into an error where `metabase/lib/react-compat`'s `isReact17OrEarlier` was failing. The check was running 
```
export const isReact17OrEarlier = () => !("useSyncExternalStore" in React);
```

However, in `embedding-sdk/lib/polyfill-use-sync-external-store.ts` we've already polyfilled `useSyncExternalStore`, so the check would always return false.


_**Fix**_

I've changed the check to use `React.version`, which should tell us the react version more reliably than checking for a key within the `React` object. I've also added unit tests for this function.

```
export const isReactVersionLessThanOrEqualTo17 = (version: string) => {
  const versionParts = version.split(".").map(Number);
  return versionParts[0] < 17 || versionParts[0] === 17;
};
```

_**How to reproduce**_

* Pull this repo into your local instance, and create a question with a table visualization. Use the 'info' button to open the sidesheet, and copy the Entity ID.
* Use [this repro](https://github.com/oisincoveney/repro-lifesight) to instantiate a Metabase Embedding SDK app with React 17 and `craco`
* Ensure that the reference to `@metabase/embedding-sdk-react` is pointed to your repo, and you've added your API key in the `Provider` component
* Run the React app. The table should appear without errors.
